In [4]:
#import packages
import pandas as pd
from tqdm import tqdm
import pickle
import numpy as np

## 1. Import data from Spotify and Pitchfork

In [186]:
pitchfork_df = pd.read_pickle('pitchfork_df.pickle')
features_df = pd.read_pickle('song_features_df.pkl')
objects_df = pd.read_pickle('song_objects_df.pkl')

In [187]:
pitchfork_df.head()

Album           Artist Score       Genres  \
0             Petals for Armor  Hayley Williams   7.2  Pop/R&BRock   
1                      Regresa       Buscabulla   7.7   Electronic   
2          What’s New, Tomboy?    Damien Jurado   7.3         Rock   
3                  Shadow Talk       Cafe Racer   7.4         Rock   
4  It Was Good Until It Wasn’t          Kehlani   7.7      Pop/R&B   

                                             Summary  
0  The Paramore singer’s debut solo album is emot...  
1  The Puerto Rican duo moved back to the island ...  
2  Channeling stripped-back pop songs into hushed...  
3  The Chicago band’s third album brings wiry new...  
4  The cloudy grooves of the Oakland singer’s sec...

In [188]:
features_df.head()

danceability  energy   key  loudness  mode  speechiness  acousticness  \
0         0.790   0.599     7    -7.417     1       0.0431         0.218   
1         0.570   0.339     2   -11.857     0       0.0449         0.569   
2           NaN     NaN  None       NaN  None          NaN           NaN   
3         0.877   0.598     6    -6.292     0       0.0296         0.353   
4         0.564   0.391     7    -7.816     1       0.0785         0.235   

   instrumentalness  liveness  valence    tempo time_signature  \
0          0.002570     0.118    0.362  109.978              4   
1          0.451000     0.126    0.442   88.963              4   
2               NaN       NaN      NaN      NaN           None   
3          0.001320     0.380    0.671   97.012              4   
4          0.000054     0.110    0.419  101.449              4   

                  song_id  
0  2moHnkHTSXBe9KjvPSQJvg  
1  5ccn0VZg8HQZsZXxGx7UZQ  
2                    None  
3  5ZD0jxyhmdNZ3DwDRXqV5U  
4  1y2GDXFvtHPbCkHFYg49Gn

In [189]:
objects_df.head()

name duration_ms popularity num_markets             album  \
0          Simmer      266115         64          79  Petals For Armor   
1  Leave It Alone      245076         61          79  Petals For Armor   
2        Cinnamon      211834         61          79  Petals For Armor   
3        Creepin'      178398         60          79  Petals For Armor   
4   Sudden Desire      187969         60          79  Petals For Armor   

  disc_number is_explicit track_number release_date           artist  \
0           1        True            1   2020-05-08  Hayley Williams   
1           1       False            2   2020-05-08  Hayley Williams   
2           1       False            3   2020-05-08  Hayley Williams   
3           1       False            4   2020-05-08  Hayley Williams   
4           1       False            5   2020-05-08  Hayley Williams   

                  song_id                album_id  
0  2moHnkHTSXBe9KjvPSQJvg  4HXpQ5KQBVWN25ltjnX7xa  
1  5ccn0VZg8HQZsZXxGx7UZQ  4HXpQ5KQBVWN25ltjnX7xa  
2  48pL4zB6KXWPvd7Ln33ENy  4HXpQ5KQBVWN25ltjnX7xa  
3  5ZD0jxyhmdNZ3DwDRXqV5U  4HXpQ5KQBVWN25ltjnX7xa  
4  1y2GDXFvtHPbCkHFYg49Gn  4HXpQ5KQBVWN25ltjnX7xa

## 2. Merge Spotify objects and features dataframes

In [190]:
df_spotify = pd.merge(features_df, objects_df, on='song_id')

In [191]:
#in order to see all the columns
pd.set_option('display.max_columns', None)

In [192]:
df_spotify.head()

danceability  energy   key  loudness  mode  speechiness  acousticness  \
0          0.79   0.599     7    -7.417     1       0.0431         0.218   
1          0.57   0.339     2   -11.857     0       0.0449         0.569   
2           NaN     NaN  None       NaN  None          NaN           NaN   
3           NaN     NaN  None       NaN  None          NaN           NaN   
4           NaN     NaN  None       NaN  None          NaN           NaN   

   instrumentalness  liveness  valence    tempo time_signature  \
0           0.00257     0.118    0.362  109.978              4   
1           0.45100     0.126    0.442   88.963              4   
2               NaN       NaN      NaN      NaN           None   
3               NaN       NaN      NaN      NaN           None   
4               NaN       NaN      NaN      NaN           None   

                  song_id            name duration_ms popularity num_markets  \
0  2moHnkHTSXBe9KjvPSQJvg          Simmer      266115         64          79   
1  5ccn0VZg8HQZsZXxGx7UZQ  Leave It Alone      245076         61          79   
2                    None            None        None       None        None   
3                    None            None        None       None        None   
4                    None            None        None       None        None   

              album disc_number is_explicit track_number release_date  \
0  Petals For Armor           1        True            1   2020-05-08   
1  Petals For Armor           1       False            2   2020-05-08   
2              None        None        None         None         None   
3              None        None        None         None         None   
4              None        None        None         None         None   

            artist                album_id  
0  Hayley Williams  4HXpQ5KQBVWN25ltjnX7xa  
1  Hayley Williams  4HXpQ5KQBVWN25ltjnX7xa  
2             None                    None  
3             None                    None  
4             None                    None

In [193]:
df_spotify.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 855902 entries, 0 to 855901
Data columns (total 24 columns):
danceability        23738 non-null float64
energy              23738 non-null float64
key                 23738 non-null object
loudness            23738 non-null float64
mode                23738 non-null object
speechiness         23738 non-null float64
acousticness        23738 non-null float64
instrumentalness    23738 non-null float64
liveness            23738 non-null float64
valence             23738 non-null float64
tempo               23738 non-null float64
time_signature      23738 non-null object
song_id             23738 non-null object
name                23738 non-null object
duration_ms         23738 non-null object
popularity          23738 non-null object
num_markets         23738 non-null object
album               23738 non-null object
disc_number         23738 non-null object
is_explicit         23738 non-null object
track_number        23738 non-null objec

In [194]:
#remove albums with no information
df_spotify = df_spotify[df_spotify['energy'].notna()]

In [195]:
df_spotify.head()

danceability  energy key  loudness mode  speechiness  acousticness  \
0              0.790   0.599   7    -7.417    1       0.0431         0.218   
1              0.570   0.339   2   -11.857    0       0.0449         0.569   
832166         0.877   0.598   6    -6.292    0       0.0296         0.353   
832167         0.564   0.391   7    -7.816    1       0.0785         0.235   
832168         0.737   0.787   7    -4.587    1       0.0525         0.121   

        instrumentalness  liveness  valence    tempo time_signature  \
0               0.002570     0.118    0.362  109.978              4   
1               0.451000     0.126    0.442   88.963              4   
832166          0.001320     0.380    0.671   97.012              4   
832167          0.000054     0.110    0.419  101.449              4   
832168          0.000002     0.114    0.590   90.014              4   

                       song_id            name duration_ms popularity  \
0       2moHnkHTSXBe9KjvPSQJvg          Simmer      266115         64   
1       5ccn0VZg8HQZsZXxGx7UZQ  Leave It Alone      245076         61   
832166  5ZD0jxyhmdNZ3DwDRXqV5U        Creepin'      178398         60   
832167  1y2GDXFvtHPbCkHFYg49Gn   Sudden Desire      187969         60   
832168  5RdkyZw1KQ1v0EbquUAsx2      Dead Horse      199412         64   

       num_markets             album disc_number is_explicit track_number  \
0               79  Petals For Armor           1        True            1   
1               79  Petals For Armor           1       False            2   
832166          79  Petals For Armor           1       False            4   
832167          79  Petals For Armor           1       False            5   
832168          79  Petals For Armor           2       False            1   

       release_date           artist                album_id  
0        2020-05-08  Hayley Williams  4HXpQ5KQBVWN25ltjnX7xa  
1        2020-05-08  Hayley Williams  4HXpQ5KQBVWN25ltjnX7xa  
832166   2020-05-08  Hayley Williams  4HXpQ5KQBVWN25ltjnX7xa  
832167   2020-05-08  Hayley Williams  4HXpQ5KQBVWN25ltjnX7xa  
832168   2020-05-08  Hayley Williams  4HXpQ5KQBVWN25ltjnX7xa

In [196]:
df_spotify.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 23738 entries, 0 to 855901
Data columns (total 24 columns):
danceability        23738 non-null float64
energy              23738 non-null float64
key                 23738 non-null object
loudness            23738 non-null float64
mode                23738 non-null object
speechiness         23738 non-null float64
acousticness        23738 non-null float64
instrumentalness    23738 non-null float64
liveness            23738 non-null float64
valence             23738 non-null float64
tempo               23738 non-null float64
time_signature      23738 non-null object
song_id             23738 non-null object
name                23738 non-null object
duration_ms         23738 non-null object
popularity          23738 non-null object
num_markets         23738 non-null object
album               23738 non-null object
disc_number         23738 non-null object
is_explicit         23738 non-null object
track_number        23738 non-null object

## 3. Merge Spotify and Pitchfork info

In [197]:
#apply lower case to albums and titles for merging
pitchfork_df['Album'] = pitchfork_df['Album'].astype(str).apply(lambda x: x.lower())
pitchfork_df['Artist'] = pitchfork_df['Artist'].astype(str).apply(lambda x: x.lower())

In [198]:
pitchfork_df.head()

Album           Artist Score       Genres  \
0             petals for armor  hayley williams   7.2  Pop/R&BRock   
1                      regresa       buscabulla   7.7   Electronic   
2          what’s new, tomboy?    damien jurado   7.3         Rock   
3                  shadow talk       cafe racer   7.4         Rock   
4  it was good until it wasn’t          kehlani   7.7      Pop/R&B   

                                             Summary  
0  The Paramore singer’s debut solo album is emot...  
1  The Puerto Rican duo moved back to the island ...  
2  Channeling stripped-back pop songs into hushed...  
3  The Chicago band’s third album brings wiry new...  
4  The cloudy grooves of the Oakland singer’s sec...

In [199]:
#rename columns to match for merging
df_spotify.rename(columns={"album": "Album", "artist": "Artist"}, inplace=True)

In [200]:
#apply lower case to albums and titles for merging
df_spotify['Album'] = df_spotify['Album'].astype(str).apply(lambda x: x.lower())
df_spotify['Artist'] = df_spotify['Artist'].astype(str).apply(lambda x: x.lower())

In [201]:
df_spotify.head()

danceability  energy key  loudness mode  speechiness  acousticness  \
0              0.790   0.599   7    -7.417    1       0.0431         0.218   
1              0.570   0.339   2   -11.857    0       0.0449         0.569   
832166         0.877   0.598   6    -6.292    0       0.0296         0.353   
832167         0.564   0.391   7    -7.816    1       0.0785         0.235   
832168         0.737   0.787   7    -4.587    1       0.0525         0.121   

        instrumentalness  liveness  valence    tempo time_signature  \
0               0.002570     0.118    0.362  109.978              4   
1               0.451000     0.126    0.442   88.963              4   
832166          0.001320     0.380    0.671   97.012              4   
832167          0.000054     0.110    0.419  101.449              4   
832168          0.000002     0.114    0.590   90.014              4   

                       song_id            name duration_ms popularity  \
0       2moHnkHTSXBe9KjvPSQJvg          Simmer      266115         64   
1       5ccn0VZg8HQZsZXxGx7UZQ  Leave It Alone      245076         61   
832166  5ZD0jxyhmdNZ3DwDRXqV5U        Creepin'      178398         60   
832167  1y2GDXFvtHPbCkHFYg49Gn   Sudden Desire      187969         60   
832168  5RdkyZw1KQ1v0EbquUAsx2      Dead Horse      199412         64   

       num_markets             Album disc_number is_explicit track_number  \
0               79  petals for armor           1        True            1   
1               79  petals for armor           1       False            2   
832166          79  petals for armor           1       False            4   
832167          79  petals for armor           1       False            5   
832168          79  petals for armor           2       False            1   

       release_date           Artist                album_id  
0        2020-05-08  hayley williams  4HXpQ5KQBVWN25ltjnX7xa  
1        2020-05-08  hayley williams  4HXpQ5KQBVWN25ltjnX7xa  
832166   2020-05-08  hayley williams  4HXpQ5KQBVWN25ltjnX7xa  
832167   2020-05-08  hayley williams  4HXpQ5KQBVWN25ltjnX7xa  
832168   2020-05-08  hayley williams  4HXpQ5KQBVWN25ltjnX7xa

In [202]:
#merge spotify and pitchfork dataframes
df = pitchfork_df.merge(df_spotify, how='inner')

In [203]:
#dropping rows with value == null
df = df[(df != 'null').all(axis=1)]

/Users/brendonius/opt/anaconda3/lib/python3.7/site-packages/pandas/core/ops/__init__.py:1115: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = method(y)


In [204]:
df.head()

Album           Artist Score       Genres  \
0  petals for armor  hayley williams   7.2  Pop/R&BRock   
1  petals for armor  hayley williams   7.2  Pop/R&BRock   
2  petals for armor  hayley williams   7.2  Pop/R&BRock   
3  petals for armor  hayley williams   7.2  Pop/R&BRock   
4  petals for armor  hayley williams   7.2  Pop/R&BRock   

                                             Summary  danceability  energy  \
0  The Paramore singer’s debut solo album is emot...         0.790   0.599   
1  The Paramore singer’s debut solo album is emot...         0.570   0.339   
2  The Paramore singer’s debut solo album is emot...         0.877   0.598   
3  The Paramore singer’s debut solo album is emot...         0.564   0.391   
4  The Paramore singer’s debut solo album is emot...         0.737   0.787   

  key  loudness mode  speechiness  acousticness  instrumentalness  liveness  \
0   7    -7.417    1       0.0431         0.218          0.002570     0.118   
1   2   -11.857    0       0.0449         0.569          0.451000     0.126   
2   6    -6.292    0       0.0296         0.353          0.001320     0.380   
3   7    -7.816    1       0.0785         0.235          0.000054     0.110   
4   7    -4.587    1       0.0525         0.121          0.000002     0.114   

   valence    tempo time_signature                 song_id            name  \
0    0.362  109.978              4  2moHnkHTSXBe9KjvPSQJvg          Simmer   
1    0.442   88.963              4  5ccn0VZg8HQZsZXxGx7UZQ  Leave It Alone   
2    0.671   97.012              4  5ZD0jxyhmdNZ3DwDRXqV5U        Creepin'   
3    0.419  101.449              4  1y2GDXFvtHPbCkHFYg49Gn   Sudden Desire   
4    0.590   90.014              4  5RdkyZw1KQ1v0EbquUAsx2      Dead Horse   

  duration_ms popularity num_markets disc_number is_explicit track_number  \
0      266115         64          79           1        True            1   
1      245076         61          79           1       False            2   
2      178398         60          79           1       False            4   
3      187969         60          79           1       False            5   
4      199412         64          79           2       False            1   

  release_date                album_id  
0   2020-05-08  4HXpQ5KQBVWN25ltjnX7xa  
1   2020-05-08  4HXpQ5KQBVWN25ltjnX7xa  
2   2020-05-08  4HXpQ5KQBVWN25ltjnX7xa  
3   2020-05-08  4HXpQ5KQBVWN25ltjnX7xa  
4   2020-05-08  4HXpQ5KQBVWN25ltjnX7xa

In [205]:
df['album_id'].nunique()

1617

#### I now have 1,617 unique albums to work with

## 4. Create dataframe for modeling

In [206]:
#isolate desired columns
model_df = df[['Score','Genres', 'danceability', 'energy', 'key', 'loudness', 'mode', 'speechiness',
               'acousticness', 'instrumentalness', 'liveness', 'valence', 'tempo', 'time_signature',
               'song_id', 'duration_ms',  'popularity', 'num_markets', 'disc_number', 'is_explicit',
               'track_number', 'album_id']]

In [207]:
model_df.head()

Score       Genres  danceability  energy key  loudness mode  speechiness  \
0   7.2  Pop/R&BRock         0.790   0.599   7    -7.417    1       0.0431   
1   7.2  Pop/R&BRock         0.570   0.339   2   -11.857    0       0.0449   
2   7.2  Pop/R&BRock         0.877   0.598   6    -6.292    0       0.0296   
3   7.2  Pop/R&BRock         0.564   0.391   7    -7.816    1       0.0785   
4   7.2  Pop/R&BRock         0.737   0.787   7    -4.587    1       0.0525   

   acousticness  instrumentalness  liveness  valence    tempo time_signature  \
0         0.218          0.002570     0.118    0.362  109.978              4   
1         0.569          0.451000     0.126    0.442   88.963              4   
2         0.353          0.001320     0.380    0.671   97.012              4   
3         0.235          0.000054     0.110    0.419  101.449              4   
4         0.121          0.000002     0.114    0.590   90.014              4   

                  song_id duration_ms popularity num_markets disc_number  \
0  2moHnkHTSXBe9KjvPSQJvg      266115         64          79           1   
1  5ccn0VZg8HQZsZXxGx7UZQ      245076         61          79           1   
2  5ZD0jxyhmdNZ3DwDRXqV5U      178398         60          79           1   
3  1y2GDXFvtHPbCkHFYg49Gn      187969         60          79           1   
4  5RdkyZw1KQ1v0EbquUAsx2      199412         64          79           2   

  is_explicit track_number                album_id  
0        True            1  4HXpQ5KQBVWN25ltjnX7xa  
1       False            2  4HXpQ5KQBVWN25ltjnX7xa  
2       False            4  4HXpQ5KQBVWN25ltjnX7xa  
3       False            5  4HXpQ5KQBVWN25ltjnX7xa  
4       False            1  4HXpQ5KQBVWN25ltjnX7xa

In [208]:
model_df.Genres.value_counts()

Rock                                  6040
Rap                                   3935
Pop/R&B                               2446
Electronic                            2312
Experimental                          1448
Folk/Country                           850
Jazz                                   469
Metal                                  442
ElectronicRock                         375
ExperimentalRock                       104
Global                                  96
Pop/R&BRock                             96
JazzPop/R&B                             67
Pop/R&BRap                              60
ElectronicPop/R&B                       44
ElectronicExperimental                  42
RapRock                                 31
ExperimentalElectronic                  29
GlobalPop/R&B                           27
ExperimentalRap                         25
Folk/CountryPop/R&B                     23
MetalRock                               21
RockExperimental                        16
Pop/R&BElec

In [209]:
#replace subgenres with main genre
model_df['Genres'].replace(['Pop/R&BRock'], 'Pop/R&B', inplace=True)
model_df['Genres'].replace(['RockExperimental'], 'ExperimentalRock', inplace=True)
model_df['Genres'].replace(['Pop/R&BJazz'], 'Pop/R&B', inplace=True)
model_df['Genres'].replace(['RapFolk/Country'], 'Rap', inplace=True)
model_df['Genres'].replace(['ElectronicJazzRock'], 'ElectronicRock', inplace=True)
model_df['Genres'].replace(['ExperimentalPop/R&B'], 'Pop/R&B', inplace=True)
model_df['Genres'].replace(['Pop/R&BFolk/Country'], 'Pop/R&B', inplace=True)
model_df['Genres'].replace(['RockFolk/Country'], 'Rock', inplace=True)
model_df['Genres'].replace(['Pop/R&BExperimental'], 'Pop/R&B', inplace=True)
model_df['Genres'].replace(['ElectronicExperimentalFolk/Country'], 'Electronic', inplace=True)
model_df['Genres'].replace(['ElectronicRap'], 'Rap', inplace=True)
model_df['Genres'].replace(['ElectronicJazzPop/R&B'], 'Electronic', inplace=True)
model_df['Genres'].replace(['ElectronicExperimentalMetal'], 'ExperimentalElectronic', inplace=True)
model_df['Genres'].replace(['RockMetal'], 'Metal', inplace=True)
model_df['Genres'].replace(['RockRap'], 'Rock', inplace=True)
model_df['Genres'].replace(['ElectronicExperimentalRock'], 'ExperimentalElectronic', inplace=True)
model_df['Genres'].replace(['Pop/R&BElectronic'], 'Pop/R&B', inplace=True)
model_df['Genres'].replace(['ElectronicExperimental'], 'ExperimentalElectronic', inplace=True)
model_df['Genres'].replace(['RapRock'], 'Rap', inplace=True)
model_df['Genres'].replace(['ExperimentalRap'], 'Rap', inplace=True)
model_df['Genres'].replace(['Folk/CountryPop/R&B'], 'Folk/Country', inplace=True)
model_df['Genres'].replace(['GlobalPop/R&B'], 'Global', inplace=True)
model_df['Genres'].replace(['MetalRock'], 'Metal', inplace=True)
model_df['Genres'].replace(['Pop/R&BRap'], 'Pop/R&B', inplace=True)
model_df['Genres'].replace(['JazzPop/R&B'], 'Jazz', inplace=True)
model_df['Genres'].replace(['ElectronicPop/R&B'], 'Pop/R&B', inplace=True)


/Users/brendonius/opt/anaconda3/lib/python3.7/site-packages/pandas/core/generic.py:6786: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


In [210]:
model_df.Genres.value_counts()

Rock                      6062
Rap                       4010
Pop/R&B                   2699
Electronic                2333
Experimental              1448
Folk/Country               873
Jazz                       536
Metal                      474
ElectronicRock             385
Global                     123
ExperimentalRock           120
ExperimentalElectronic      94
Name: Genres, dtype: int64

In [211]:
model_df.Genres.nunique()

12

In [212]:
#change is_explicit column bool to numeric
model_df['is_explicit'].replace(to_replace='True', value = 1, inplace=True)
model_df['is_explicit'].replace(to_replace='False', value = 0, inplace=True)

In [213]:
model_df.is_explicit.head()

0     True
1    False
2    False
3    False
4    False
Name: is_explicit, dtype: bool

In [214]:
type(model_df.Score[0])

str

In [232]:
#change score column from string to float
model_df['Score'] = model_df['Score'].astype(float)
model_df['popularity'] = model_df['popularity'].astype(float)
model_df['mode'] = model_df['mode'].astype(float)
model_df['key'] = model_df['key'].astype(float)

/Users/brendonius/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/Users/brendonius/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/Users/brendonius/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataF

In [233]:
model_df.head()

Score   Genres  danceability  energy  key  loudness  mode  speechiness  \
0    7.2  Pop/R&B         0.790   0.599  7.0    -7.417   1.0       0.0431   
1    7.2  Pop/R&B         0.570   0.339  2.0   -11.857   0.0       0.0449   
2    7.2  Pop/R&B         0.877   0.598  6.0    -6.292   0.0       0.0296   
3    7.2  Pop/R&B         0.564   0.391  7.0    -7.816   1.0       0.0785   
4    7.2  Pop/R&B         0.737   0.787  7.0    -4.587   1.0       0.0525   

   acousticness  instrumentalness  liveness  valence    tempo time_signature  \
0         0.218          0.002570     0.118    0.362  109.978              4   
1         0.569          0.451000     0.126    0.442   88.963              4   
2         0.353          0.001320     0.380    0.671   97.012              4   
3         0.235          0.000054     0.110    0.419  101.449              4   
4         0.121          0.000002     0.114    0.590   90.014              4   

                  song_id duration_ms  popularity num_markets disc_number  \
0  2moHnkHTSXBe9KjvPSQJvg      266115        64.0          79           1   
1  5ccn0VZg8HQZsZXxGx7UZQ      245076        61.0          79           1   
2  5ZD0jxyhmdNZ3DwDRXqV5U      178398        60.0          79           1   
3  1y2GDXFvtHPbCkHFYg49Gn      187969        60.0          79           1   
4  5RdkyZw1KQ1v0EbquUAsx2      199412        64.0          79           2   

   is_explicit track_number                album_id  
0         True            1  4HXpQ5KQBVWN25ltjnX7xa  
1        False            2  4HXpQ5KQBVWN25ltjnX7xa  
2        False            4  4HXpQ5KQBVWN25ltjnX7xa  
3        False            5  4HXpQ5KQBVWN25ltjnX7xa  
4        False            1  4HXpQ5KQBVWN25ltjnX7xa

In [234]:
#group songs by album_id
grouped_albums = model_df.groupby('album_id')

In [235]:
grouped_albums.size()

album_id
007DWn799UWvfY1wwZeENR    15
00ojDEHvnEOPOjhHrUrKy5    10
00oz3t7cI3WfwS2oEIZD6x     9
00uqjqwLruO346z3cNlNqG    13
01ey0ETPITd3jiOiztRXzK    16
                          ..
7z4GhRfLqfSkqrj5F3Yt2B    12
7zIABwRfxyMXWDEaLm89HG    14
7zQuYvRVyQxgIlxhcSLTdW    18
7zeCZY6rQRufc8IHGKyXGX    11
7zgo8oFUtR2cVQbhDZ0o6V     6
Length: 1617, dtype: int64

In [236]:
#get mean values of numerical variables
grouped_mean_df = grouped_albums['danceability', 'energy', 'acousticness', 'instrumentalness', 
                            'liveness', 'loudness', 'speechiness', 'valence', 'tempo', 
                            'time_signature', 'popularity', 'Score', 'num_markets', 'is_explicit', 'mode', 'key'].agg([np.mean])

In [237]:
grouped_mean_df.columns = ["_".join(x) for x in grouped_mean_df.columns.ravel()]

In [238]:
grouped_mean_df.head()

danceability_mean  energy_mean  acousticness_mean  \
album_id                                                                    
007DWn799UWvfY1wwZeENR           0.871667     0.569000           0.046364   
00ojDEHvnEOPOjhHrUrKy5           0.724300     0.384600           0.374767   
00oz3t7cI3WfwS2oEIZD6x           0.443889     0.174333           0.957778   
00uqjqwLruO346z3cNlNqG           0.562154     0.601769           0.464462   
01ey0ETPITd3jiOiztRXzK           0.377625     0.494812           0.406806   

                        instrumentalness_mean  liveness_mean  loudness_mean  \
album_id                                                                      
007DWn799UWvfY1wwZeENR               0.000119       0.177453      -7.147600   
00ojDEHvnEOPOjhHrUrKy5               0.532800       0.200730     -15.015700   
00oz3t7cI3WfwS2oEIZD6x               0.205918       0.114111     -13.239333   
00uqjqwLruO346z3cNlNqG               0.011059       0.148254      -6.111769   
01ey0ETPITd3jiOiztRXzK               0.049877       0.147925     -12.121375   

                        speechiness_mean  valence_mean  tempo_mean  \
album_id                                                             
007DWn799UWvfY1wwZeENR          0.254213      0.413020  126.265133   
00ojDEHvnEOPOjhHrUrKy5          0.094240      0.362970  100.700800   
00oz3t7cI3WfwS2oEIZD6x          0.037156      0.214989  118.242111   
00uqjqwLruO346z3cNlNqG          0.084700      0.455692  124.775154   
01ey0ETPITd3jiOiztRXzK          0.040194      0.268081  105.696000   

                        popularity_mean  Score_mean  is_explicit_mean  \
album_id                                                                
007DWn799UWvfY1wwZeENR        62.000000         7.8            1.0000   
00ojDEHvnEOPOjhHrUrKy5         4.500000         7.4            0.0000   
00oz3t7cI3WfwS2oEIZD6x        33.000000         8.4            0.0000   
00uqjqwLruO346z3cNlNqG        51.461538         7.4            0.0000   
01ey0ETPITd3jiOiztRXzK         1.875000         7.1            0.0625   

                        mode_mean  key_mean  
album_id                                     
007DWn799UWvfY1wwZeENR   0.666667  4.400000  
00ojDEHvnEOPOjhHrUrKy5   0.500000  8.100000  
00oz3t7cI3WfwS2oEIZD6x   0.444444  4.777778  
00uqjqwLruO346z3cNlNqG   0.615385  5.153846  
01ey0ETPITd3jiOiztRXzK   0.687500  4.375000

In [239]:
#get album length in time and number of tracks in an album
album_durations_df = grouped_albums.agg(album_run_time_ms = pd.NamedAgg(column='duration_ms', aggfunc=np.sum),
                                        number_of_tracks = pd.NamedAgg(column='track_number', aggfunc='count')
                                       )

In [240]:
album_durations_df = album_durations_df[album_durations_df.number_of_tracks <= 35]

In [241]:
album_durations_df.head()

album_run_time_ms  number_of_tracks
album_id                                                   
007DWn799UWvfY1wwZeENR            3065169                15
00ojDEHvnEOPOjhHrUrKy5            2886000                10
00oz3t7cI3WfwS2oEIZD6x            1670290                 9
00uqjqwLruO346z3cNlNqG            2337540                13
01ey0ETPITd3jiOiztRXzK            2840581                16

In [242]:
#concatenate the two dataframes
album_df = pd.concat([album_durations_df, grouped_mean_df], axis=1, sort=False)

In [243]:
#add genres column to dataframe
genres = grouped_albums['Genres'].first()
album_df['genres'] = genres

In [244]:
album_df.genres.value_counts()

Rock                      531
Rap                       268
Electronic                219
Pop/R&B                   213
Experimental              152
Folk/Country               82
Jazz                       49
Metal                      48
ElectronicRock             22
ExperimentalRock           12
Global                     11
ExperimentalElectronic     10
Name: genres, dtype: int64

In [245]:
album_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1617 entries, 007DWn799UWvfY1wwZeENR to 7LncrTHGjSw88ZcNOdZMYQ
Data columns (total 17 columns):
album_run_time_ms        1604 non-null float64
number_of_tracks         1604 non-null float64
danceability_mean        1617 non-null float64
energy_mean              1617 non-null float64
acousticness_mean        1617 non-null float64
instrumentalness_mean    1617 non-null float64
liveness_mean            1617 non-null float64
loudness_mean            1617 non-null float64
speechiness_mean         1617 non-null float64
valence_mean             1617 non-null float64
tempo_mean               1617 non-null float64
popularity_mean          1617 non-null float64
Score_mean               1617 non-null float64
is_explicit_mean         1617 non-null float64
mode_mean                1617 non-null float64
key_mean                 1617 non-null float64
genres                   1617 non-null object
dtypes: float64(16), object(1)
memory usage: 227.4+ KB


In [246]:
album_df.head()

album_run_time_ms  number_of_tracks  \
007DWn799UWvfY1wwZeENR          3065169.0              15.0   
00ojDEHvnEOPOjhHrUrKy5          2886000.0              10.0   
00oz3t7cI3WfwS2oEIZD6x          1670290.0               9.0   
00uqjqwLruO346z3cNlNqG          2337540.0              13.0   
01ey0ETPITd3jiOiztRXzK          2840581.0              16.0   

                        danceability_mean  energy_mean  acousticness_mean  \
007DWn799UWvfY1wwZeENR           0.871667     0.569000           0.046364   
00ojDEHvnEOPOjhHrUrKy5           0.724300     0.384600           0.374767   
00oz3t7cI3WfwS2oEIZD6x           0.443889     0.174333           0.957778   
00uqjqwLruO346z3cNlNqG           0.562154     0.601769           0.464462   
01ey0ETPITd3jiOiztRXzK           0.377625     0.494812           0.406806   

                        instrumentalness_mean  liveness_mean  loudness_mean  \
007DWn799UWvfY1wwZeENR               0.000119       0.177453      -7.147600   
00ojDEHvnEOPOjhHrUrKy5               0.532800       0.200730     -15.015700   
00oz3t7cI3WfwS2oEIZD6x               0.205918       0.114111     -13.239333   
00uqjqwLruO346z3cNlNqG               0.011059       0.148254      -6.111769   
01ey0ETPITd3jiOiztRXzK               0.049877       0.147925     -12.121375   

                        speechiness_mean  valence_mean  tempo_mean  \
007DWn799UWvfY1wwZeENR          0.254213      0.413020  126.265133   
00ojDEHvnEOPOjhHrUrKy5          0.094240      0.362970  100.700800   
00oz3t7cI3WfwS2oEIZD6x          0.037156      0.214989  118.242111   
00uqjqwLruO346z3cNlNqG          0.084700      0.455692  124.775154   
01ey0ETPITd3jiOiztRXzK          0.040194      0.268081  105.696000   

                        popularity_mean  Score_mean  is_explicit_mean  \
007DWn799UWvfY1wwZeENR        62.000000         7.8            1.0000   
00ojDEHvnEOPOjhHrUrKy5         4.500000         7.4            0.0000   
00oz3t7cI3WfwS2oEIZD6x        33.000000         8.4            0.0000   
00uqjqwLruO346z3cNlNqG        51.461538         7.4            0.0000   
01ey0ETPITd3jiOiztRXzK         1.875000         7.1            0.0625   

                        mode_mean  key_mean          genres  
007DWn799UWvfY1wwZeENR   0.666667  4.400000             Rap  
00ojDEHvnEOPOjhHrUrKy5   0.500000  8.100000      Electronic  
00oz3t7cI3WfwS2oEIZD6x   0.444444  4.777778    Folk/Country  
00uqjqwLruO346z3cNlNqG   0.615385  5.153846         Pop/R&B  
01ey0ETPITd3jiOiztRXzK   0.687500  4.375000  ElectronicRock

In [247]:
album_df = album_df[album_df['number_of_tracks'].notna()]

In [248]:
#save df for modeling
with open('album_df.pkl', 'wb') as f:
    pickle.dump(album_df, f)